File: tier2_intermediate/dfs/problems/p450_delete_node_in_a_bst.cpp
/*
 * P450: Delete Node in a BST [PREMIUM] (Medium)
 * https://leetcode.com/problems/delete-node-in-a-bst/
 * Topics: Tree, Binary Search Tree, Binary Tree
 */

#include "io.h"
#include <cstdio>
#include <queue>
#include <string>
#include <vector>

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

static TreeNode *insert(TreeNode *root, int val) {
    if (!root) return new TreeNode(val);
    if (val < root->val) root->left = insert(root->left, val);
    else root->right = insert(root->right, val);
    return root;
}

static TreeNode *deleteNode(TreeNode *root, int key) {
    if (!root) return nullptr;
    if (key < root->val) root->left = deleteNode(root->left, key);
    else if (key > root->val) root->right = deleteNode(root->right, key);
    else {
        if (!root->left) { auto *t = root->right; delete root; return t; }
        if (!root->right) { auto *t = root->left; delete root; return t; }
        TreeNode *mn = root->right;
        while (mn->left) mn = mn->left;
        root->val = mn->val;
        root->right = deleteNode(root->right, mn->val);
    }
    return root;
}

#define NL 2147483647

int main() {
    abort();
}

File: tier2_intermediate/dfs/problems/p450_delete_node_in_a_bst.rs
/*
 * P450: Delete Node in a BST [PREMIUM] (Medium)
 * https://leetcode.com/problems/delete-node-in-a-bst/
 * Topics: Tree, Binary Search Tree, Binary Tree
 */

use wasm_libs::*;
use std::io::{self, Write};

#[derive(Debug, Clone)]
struct TreeNode {
    val: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

fn insert(root: Option<Box<TreeNode>>, val: i32) -> Option<Box<TreeNode>> {
    todo!();
}

fn find_min(node: &Box<TreeNode>) -> i32 {
    todo!();
}

fn delete_node(root: Option<Box<TreeNode>>, key: i32) -> Option<Box<TreeNode>> {
    todo!();
}

fn main() {
    let tree_line = read_line();
    let key: i32 = read_line().trim().parse().unwrap();
    let nl = "null";
    let mut root: Option<Box<TreeNode>> = None;
    for tok in tree_line.split_whitespace() {
        if tok == nl { continue; }
        let v: i32 = tok.parse().unwrap();
        root = insert(root, v);
    }
    let root = delete_node(root, key);
    match root {
        None => { let mut out = io::stdout().lock(); writeln!(out, "null").unwrap(); },
        Some(r) => {
            let mut q = std::collections::VecDeque::new();
            q.push_back(Some(r));
            let mut first = true;
            let mut all: Vec<Option<Box<TreeNode>>> = Vec::new();
            while let Some(node) = q.pop_front() {
                all.push(node);
                if let Some(n) = all.last().unwrap() {
                    q.push_back(n.left.clone());
                    q.push_back(n.right.clone());
                }
            }
            while all.last().is_some() && all.last().unwrap().is_none() { all.pop(); }
            let mut out = io::stdout().lock();
            let mut first = true;
            for n in &all {
                if !first { write!(out, " ").unwrap(); }
                first = false;
                match n {
                    None => write!(out, "null").unwrap(),
                    Some(node) => write!(out, "{}", node.val).unwrap(),
                }
            }
            writeln!(out).unwrap();
        }
    }
}
